server/src/test/java/org/elasticsearch/cluster/routing/allocation/decider/NodeReplacementAllocationDeciderTests.java

```diff
@@ -0,0 +1,248 @@
+        RoutingAllocation allocation = new RoutingAllocation(allocationDeciders, state.getRoutingNodes(), state, null, null, 0);
+        RoutingNode routingNode = new RoutingNode(NODE_A.getId(), NODE_A, shard);
+        allocation.debugDecision(true);
+
+        Decision decision = decider.canAllocate(indexMetadata, routingNode, allocation);
+        assertThat(decision.type(), equalTo(Decision.Type.YES));
+        assertThat(decision.getExplanation(), equalTo("node is not a replacement target, so allocation is allowed"));
+
+        routingNode = new RoutingNode(NODE_B.getId(), NODE_B, shard);
+
+        decision = decider.canAllocate(indexMetadata, routingNode, allocation);
+        assertThat(decision.type(), equalTo(Decision.Type.NO));
+        assertThat(decision.getExplanation(), equalTo("node [" + NODE_B.getId() +
+            "] is replacing a vacating node, so no data from other nodes may be allocated to it until the replacement is complete"));
+
+        routingNode = new RoutingNode(NODE_C.getId(), NODE_C, shard);
+
+        decision = decider.canAllocate(indexMetadata, routingNode, allocation);
+        assertThat(decision.type(), equalTo(Decision.Type.YES));
+        assertThat(decision.getExplanation(), equalTo("node is not a replacement target, so allocation is allowed"));
+    }
+
+    public void testCannotRemainOnReplacedNode() {
+        ClusterState state = prepareState(service.reroute(ClusterState.EMPTY_STATE, "initial state"), NODE_A.getId(), NODE_B.getName());
+        RoutingAllocation allocation = new RoutingAllocation(allocationDeciders, state.getRoutingNodes(), state, null, null, 0);
+        RoutingNode routingNode = new RoutingNode(NODE_A.getId(), NODE_A, shard);
+        allocation.debugDecision(true);
+
+        Decision decision = decider.canRemain(shard, routingNode, allocation);
+        assertThat(decision.type(), equalTo(Decision.Type.NO));
+        assertThat(
+            decision.getExplanation(),
+            equalTo("node [" + NODE_A.getId() + "] is being replaced by node [" + NODE_B.getId() + "], so no data may remain on it")
+        );
+
+        routingNode = new RoutingNode(NODE_B.getId(), NODE_B, shard);
+
+        decision = decider.canRemain(shard, routingNode, allocation);
+        assertThat(decision.type(), equalTo(Decision.Type.YES));
+        assertThat(decision.getExplanation(), equalTo("node [" + NODE_B.getId() + "] is not being replaced"));
+    }
+
+    public void testCanAllocateToNeitherSourceNorTarget() {
+        ClusterState state = prepareState(service.reroute(ClusterState.EMPTY_STATE, "initial state"), NODE_A.getId(), NODE_B.getName());
+        RoutingAllocation allocation = new RoutingAllocation(allocationDeciders, state.getRoutingNodes(), state, null, null, 0);
+        RoutingNode routingNode = new RoutingNode(NODE_A.getId(), NODE_A, shard);
+        allocation.debugDecision(true);
+
+        Decision decision = decider.canAllocate(shard, routingNode, allocation);
```

Comment:
Can we check this with a shard having currentNodeId=NODE_C too? Could be just randomly picking one or the other.